### PR TITLE
Allow the same chaincode to be used on multiple channels

### PR DESCRIPTION
**Pull request by AurelienGasser**
Companion PR: 
- https://github.com/SubstraFoundation/substra-tests/pull/166

We currently don't support using the same chaincode on different channels.

This PR removes this limitation.

Sample use case: solo channels using the same chaincode as shared channels.

### CODE REVIEW

**Comment by AurelienGasser** [comment]
@Kelvin-M Could you please have a look at the overall design and let me know if that looks OK with you?

---

End-to-end tests pass, but I'm not sure the design I chose is 100% correct. In particular, I'm not sure which of the `chaincodes[]` properties should be moved to `appChannels[].chaincodes[]`

```yaml
chaincodes:
  - name: mycc
    address: network-org-1-peer-1-hlf-k8s-chaincode-mycc.org-1
    port: 7052
    policy: "OR('MyOrg1MSP.member','MyOrg2MSP.member')"
    version: "1.0"
    image:
      repository: substrafoundation/substra-chaincode
      tag: 0.1.0
      pullPolicy: IfNotPresent

appChannels:
  - name: mychannel
    chaincodes: [mycc]
```

I'm confident that the `address`, `port` and `image` properties are independent from channels. So they should be properties of `chaincodes[]`.

On the other hand the `policy` property should probably be moved to `appChannels[].chaincodes[]`. In other words, change from

```yaml
chaincodes: [mycc]
```

to 

```yaml
chaincodes:
  - name: mycc
    policy: "OR('MyOrg1MSP.member','MyOrg2MSP.member','MyOrg3MSP.member','MyOrg4MSP.member')"
```

Does that make sense?

**Comment by Kelvin-M** [comment]
As you said policy should be moved as it's the chaincode policy inside a channel and not a chaincode property :)

For the address and port, I'm not confident as you are ! If you use the same chaincode mycc for both channel, does it work ? 
I suppose, yes because each channel has a ledger, and chaincode is just applying smartcontracts over it, but I need confirmation.

**Comment by Kelvin-M** [comment]
Okay, sharing the same chaincode between to channel works ! But it means also logs sharing. So it can be difficult to understand what happen on what channel :)

**Comment by Kelvin-M** [comment]
Adding the info here : https://github.com/SubstraFoundation/substra-chaincode/pull/130/files

**Comment by AurelienGasser** [comment]
@Kelvin-M I made the policy fix.

As the new changes complicate the code a little bit, I think this PR would be a good opportunity to move the "chaincode" part of the appchannel operator to the chaincode operator. What do you think?

**Comment by Kelvin-M** [comment]
Moving "chaincode" part from the appchannel operator is not totally possible, or you will have to change the code.

If you read the Hyperledger Fabric documentation about the new chaincode lifecycle:
- https://hyperledger-fabric.readthedocs.io/en/release-2.2/chaincode_lifecycle.html
- https://hyperledger-fabric.readthedocs.io/en/release-2.2/commands/peerlifecycle.html

The chaincode deployment scenario is the following : 

```
1. Package the chaincode # chaincode operator as it depends on the Org itself
2. Install the chaincode # chaincode operator as it depends on the Org itself
3. Approve the chaincode # chaincode operator as it depends on the Org itself
4. Check chaincode commit readiness # appchannel operator as it depends on multiple Orgs
5. Commit chaincode # appchannel operator as it depends on multiple Orgs
6. Init Chaincode # appchannel operator
```
I think step 6. can be moved to chaincode operator if you want.

But for step 4. and 5., it's more complicated. In order to commit chaincode, we need to wait until it's commit readiness with `checkcommitreadiness`, explained in the documentation : 

>  You can use this command to learn whether enough channel members have approved a chaincode definition to meet the Application/Channel/Endorsement policy (a majority by default) before the definition can be committed to a channel.

Consequently, step 5. and 6. are linked and it's better to keep them together.

So why, 5. and 6. are in the appchannel operator ? Once a chaincode definition is approved by a majority of Organizations in the channel, we can commit it.

Still in the documentation :

> Once a sufficient number of organizations approve a chaincode definition for their organizations (a majority by default), one organization can commit the definition the channel using the `peer lifecycle chaincode commit` command:
This command needs to **target the peers of other organizations** on the channel to collect their organization endorsement for the definition.

We need to target the peers of other organizations in the channel to collect their endorsement of the chaincode definition.
So when we use the `peer lifecycle chaincode commit` we need to pass `--peerAddresses` and `--tlsRootCertFiles`  as showed [here](https://hyperledger-fabric.readthedocs.io/en/release-2.2/commands/peerlifecycle.html#peer-lifecycle-chaincode-commit-example) without TLS because they don't use MutualTLS like we are.

This is done [here](https://github.com/SubstraFoundation/hlf-k8s/blob/master/charts/hlf-k8s/templates/deployment-appchannel-operator.yaml#L330-L344) by using `ENDORSEMENT` env variable which is built with the peer adresses and tls root cert in the appchannel operator [here](https://github.com/SubstraFoundation/hlf-k8s/blob/master/charts/hlf-k8s/templates/deployment-appchannel-operator.yaml#L181-L184)

So if you want to move 5. and 6. in the chaincode operator you will have to rebuild the `ENDORSEMENT` variable and not using the appchannel operator way, but by using the discovery command that allows you to see which peers are connected and get their addresses and tls root cert with this [command](https://hyperledger-fabric.readthedocs.io/en/release-2.2/discovery-cli.html#peer-membership-query) or this [one](https://hyperledger-fabric.readthedocs.io/en/release-2.2/discovery-cli.html#configuration-query) 

Step 6. was put in appchannel operator for simplicity as we can init the chaincode only when it has been committed.

**Comment by AurelienGasser** [comment]
@Kelvin-M Thanks for the explanation. It's a bit more work than I thought, so I'm going to leave it out of this PR.

File: charts/hlf-k8s/Chart.yaml
Comment (ClementGautier) [inline charts/hlf-k8s/Chart.yaml:18]: That's a breaking change, you can bump to 5.x

File: charts/hlf-k8s/UPGRADE.md
Comment (ClementGautier) [inline charts/hlf-k8s/UPGRADE.md:16]: Make a 5.x section (you probably shouldn't alter a changelog)

File: charts/hlf-k8s/README.md
Comment (ClementGautier) [inline charts/hlf-k8s/README.md:132]: isn't the version necessary for the instantiation in the channel ? I'm not completely sure how it works with the lifecycle management introduced with hlf 2.x but on 1.4 you needed to pass the chaincode version during the `instatiate` command.

**Review by ClementGautier — CHANGES_REQUESTED** [review]

**Review by Kelvin-M — CHANGES_REQUESTED** [review]
Values should be updated in examples :)
If https://github.com/SubstraFoundation/substra-chaincode/pull/130 is merged, we will have to update chaincode image tag value. It can be done after.

**Comment by Kelvin-M** [comment]
Chaincode version : https://github.com/SubstraFoundation/substra-chaincode/releases/tag/0.1.1

**Comment by AurelienGasser** [comment]
@Kelvin-M which examples are you referring to?

File: charts/hlf-k8s/UPGRADE.md
Comment (AurelienGasser) [reply charts/hlf-k8s/UPGRADE.md:16]: I chose this approach for simplicity (mostly for you!) but up to you

File: charts/hlf-k8s/README.md
Comment (AurelienGasser) [reply charts/hlf-k8s/README.md:132]: The version number is still specified, although the field has been moved from `appChannels[].chaincodes[].version` to `chaincodes[].version`

**Comment by Kelvin-M** [comment]
@AurelienGasser those examples https://github.com/SubstraFoundation/hlf-k8s/tree/master/examples

File: charts/hlf-k8s/README.md
Comment (ClementGautier) [reply charts/hlf-k8s/README.md:132]: yes but that's not what I meant: in 1.4 for example you could install version x, y and z and choose to instantiate the version x in the channel A and y in the channel B. It doesn't seems possible anymore with this structure.

Comment (AurelienGasser) [reply charts/hlf-k8s/README.md:132]: Good point! 

We should be able to replicate this behavior by creating multiple chaincodes 

```
chaincodes:
- name: mycc-v1
  version: 1.0
- name: mycc-v2
  version: 2.0

appChannels:
- name: mychannel
  chaincodes: 
  - name: mycc-v1
- name: yourchannel
  chaincodes: 
  - name: mycc-v2
```

Although we haven't tested it yet.

Comment (ClementGautier) [reply charts/hlf-k8s/README.md:132]: As I understand here: https://hyperledger-fabric.readthedocs.io/en/release-2.2/chaincode_lifecycle.html#the-channel-does-not-agree-on-a-chaincode-definition it seems the chaincode commit phase still requires a pair chaincode name + version. I think you will need to put the version in both chaincodes[] (for chaincode packaging and installation) and appChannels[].chaincodes[] (for approving the definition and commiting to the channel).

Comment (Kelvin-M) [reply charts/hlf-k8s/README.md:132]: Yes, it's done [here](https://github.com/SubstraFoundation/hlf-k8s/pull/102/files#diff-93117691d56deeeb90f61349a7817cb5217d0b31080c427fecb3c683e718014dR342-R346) with chaincode properties fetch [here](https://github.com/SubstraFoundation/hlf-k8s/pull/102/files#diff-93117691d56deeeb90f61349a7817cb5217d0b31080c427fecb3c683e718014dR292-R302) @ClementGautier 

For information we use this behavior for the chaincode definition : https://hyperledger-fabric.readthedocs.io/en/release-2.2/chaincode_lifecycle.html#organizations-install-different-chaincode-packages

Comment (ClementGautier) [reply charts/hlf-k8s/README.md:132]: @Kelvin-M exactly, what I'm saying is that you shouldn't fetch the version in the chaincodes[] array based on the name of the chaincode name because that basicly forbid you to have multiple version of a same chaincode and make you use unecessary workaround like the one suggested by @AurelienGasser (putting the version in the name itself). A chaincode upgrade should be as easy as adding a new version of a chaincode in the chaincodes[] list and changing the one used in a channel by changing it's version in the appChannel.chaincode.version in my opinion. Let's discuss this on vocal this afternoon?

**Review by Kelvin-M — APPROVED** [review]
LGTM ! 
Wait for @ClementGautier approval if all its questions are answered !

**Comment by AurelienGasser** [comment]
e2e tests: :heavy_check_mark:

**Comment by AurelienGasser** [comment]
@Kelvin-M @ClementGautier The refactoring of chaincode/appchannel operators discussed earlier is addressed in https://github.com/SubstraFoundation/hlf-k8s/pull/103

File: charts/hlf-k8s/README.md
Comment (ClementGautier) [inline charts/hlf-k8s/README.md:48]: this one remains

File: charts/hlf-k8s/UPGRADE.md
Comment (ClementGautier) [inline charts/hlf-k8s/UPGRADE.md:6]: and version

**Review by ClementGautier — APPROVED** [review]
small typos to be fixed but you get my approval

**Review by Kelvin-M — APPROVED** [review]

**Review by Kelvin-M — APPROVED** [review]

**Review by ClementGautier — APPROVED** [review]

**Pull request merged**
